circuit_knitting_toolbox/circuit_cutting/cutting_evaluation.py

```diff
@@ -82,3 +183,185 @@ def append_measurement_circuit(
+            )
+        if len(circ.cregs) > 1 and circ.cregs[-2].name != "qpd_measurements":
+            raise ValueError(
+                "Circuit's next-to-last register is expected to be named 'qpd_measurements'."
+            )
+
+        if len(circ.cregs) < 2:
+            num_qpd_bits_flat.append(0)  # No gates decomposed to a measurement
+        else:
+            num_qpd_bits_flat.append(len(circ.cregs[-2]))
+
+    # Run all of the batched experiments
+    counts_flat = backend.run(experiments_flat, shots=shots).result().get_counts()
+
+    # Reshape the output data to match the input
+    counts_reshaped = np.reshape(counts_flat, np.shape(subexperiments))
+    num_qpd_bits = np.reshape(num_qpd_bits_flat, np.shape(subexperiments))
+
+    # Create the counts tuples, which include the number of QPD measurement bits
+    counts = [[] for _ in range(len(subexperiments))]
+    for i, sample in enumerate(counts_reshaped):
+        for j, count_dict in enumerate(sample):
+            counts[i].append((count_dict, num_qpd_bits[i][j]))
+
+    return counts
+
+
+def _get_mapping_ids_by_partition(
+    circuits: Sequence[QuantumCircuit],
+) -> tuple[list[list[list[int]]], list[list[int]]]:
+    """Get indices to the QPD gates in each subcircuit and relevant map ids."""
+    # Collect QPDGate id's and relevant map id's for each subcircuit
+    subcirc_qpd_gate_ids: list[list[list[int]]] = []
+    subcirc_map_ids: list[list[int]] = []
+    decomp_ids = set()
+    for circ in circuits:
+        subcirc_qpd_gate_ids.append([])
+        subcirc_map_ids.append([])
+        for i, inst in enumerate(circ.data):
+            if isinstance(inst.operation, SingleQubitQPDGate):
+                decomp_id = int(inst.operation.label.split("_")[-1])
+                decomp_ids.add(decomp_id)
+                subcirc_qpd_gate_ids[-1].append([i])
+                subcirc_map_ids[-1].append(decomp_id)
+
+    return subcirc_qpd_gate_ids, subcirc_map_ids
+
+
+def _get_bases_by_partition(
+    circuits: Sequence[QuantumCircuit], subcirc_qpd_gate_ids: list[list[list[int]]]
+) -> list[QPDBasis]:
+    """Get a list of each unique QPD basis across the subcircuits."""
+    # Collect the bases corresponding to each decomposed operation
+    bases_dict = {}
+    for i, subcirc in enumerate(subcirc_qpd_gate_ids):
+        for basis_id in subcirc:
+            decomp_id = int(
+                circuits[i].data[basis_id[0]].operation.label.split("_")[-1]
+            )
+            bases_dict[decomp_id] = circuits[i].data[basis_id[0]].operation.basis
+    bases = [bases_dict[key] for key in sorted(bases_dict.keys())]
+
+    return bases
+
+
+def _get_bases(circuit: QuantumCircuit) -> tuple[list[QPDBasis], list[list[int]]]:
+    """Get a list of each unique QPD basis in the circuit and the QPDGate indices."""
+    bases = []
+    qpd_gate_ids = []
+    for i, inst in enumerate(circuit):
+        if isinstance(inst.operation, SingleQubitQPDGate):
+            raise ValueError(
```

Comment:
I think this will still be valid for wire cutting. I dont believe there is a use case I'm aware of for wire cutting that doesn't involve separating the circuit